pkg/operator/resource/resourceapply/admissions.go

```diff
@@ -2,76 +2,159 @@ package resourceapply
+	toWrite := existingCopy // shallow copy so the code reads easier
 
-	if klog.V(4).Enabled() {
-		klog.Infof("ValidatingWebhookConfiguration %q changes: %v", required.Name, JSONPatchNoError(existing, required))
+	// Providing upgrade compatibility with service-ca-bundle operator
+	// and ignore clientConfig.caBundle changes on "inject-cabundle" label
+	if required.GetAnnotations() != nil && required.GetAnnotations()[genericCABundleInjectorAnnotation] != "" {
+		copyMutatingWebhookCABundle(existing, required)
 	}
+	toWrite.Webhooks = required.Webhooks
+
+	klog.V(4).Infof("MutatingWebhookConfiguration %q changes: %v", required.GetNamespace()+"/"+required.GetName(), JSONPatchNoError(existing, toWrite))
 
-	actual, err := client.ValidatingWebhookConfigurations().Update(context.TODO(), required, metav1.UpdateOptions{})
+	actual, err := client.MutatingWebhookConfigurations().Update(context.TODO(), toWrite, metav1.UpdateOptions{})
 	reportUpdateEvent(recorder, required, err)
-	return actual, true, err
+	if err != nil {
+		return nil, false, err
+	}
+	return actual, *modified || actual.GetGeneration() > existingCopy.GetGeneration(), nil
 }
 
-// ApplyMutatingWebhookConfiguration merges objectmeta, update webhooks.
-func ApplyMutatingWebhookConfiguration(client admissionclientv1.MutatingWebhookConfigurationsGetter, recorder events.Recorder, required *admissionv1.MutatingWebhookConfiguration) (*admissionv1.MutatingWebhookConfiguration, bool, error) {
-	existing, err := client.MutatingWebhookConfigurations().Get(context.TODO(), required.Name, metav1.GetOptions{})
+// copyMutatingWebhookCABundle populates webhooks[].clientConfig.caBundle fields from existing resource
+func copyMutatingWebhookCABundle(existing, required *admissionregistrationv1.MutatingWebhookConfiguration) {
+	existingMutatingWebhooks := make(map[string]admissionregistrationv1.MutatingWebhook)
+	for _, mutatingWebhook := range existing.Webhooks {
+		existingMutatingWebhooks[mutatingWebhook.Name] = mutatingWebhook
+	}
+
+	webhooks := make([]admissionregistrationv1.MutatingWebhook, len(required.Webhooks))
+	for i, mutatingWebhook := range required.Webhooks {
+		if webhook, ok := existingMutatingWebhooks[mutatingWebhook.Name]; ok {
+			mutatingWebhook.ClientConfig.CABundle = webhook.ClientConfig.CABundle
+		}
+		webhooks[i] = mutatingWebhook
+	}
+	required.Webhooks = webhooks
+}
+
+// ApplyValidatingWebhookConfiguration ensures the form of the specified
+// validatingwebhookconfiguration is present in the API. If it does not exist,
+// it will be created. If it does exist, the metadata of the required
+// validatingwebhookconfiguration will be merged with the existing validatingwebhookconfiguration
+// and an update performed if the validatingwebhookconfiguration spec and metadata differ from
+// the previously required spec and metadata based on generation change.
+//
+// expectedGeneration set to 0 disables generation tracking, forcing redeployment logic.
```

Comment:
what is "forcing redeployment logic" ?

Reply:
This signature breaks every other user of the func. Do we want that? `ApplyValidatingWebhookConfigurationWithGeneration` would avoid that.

Reply:
Let's break consumers IMO. According to @jsafrane there is no other consumer so.

Reply:
This is used in this PR in `ApplyDirectly`. Maybe it should not be here in the first place, I'm not sure about a use-case either. But this comment is nothing new, just highlighting the side-effect of the chosen algorithm on Deployment/DaemonSet apply logic too. For some reason, they use `forceRollout`, but it is not needed - https://github.com/openshift/library-go/blob/c8d0c55677d43fef621e8afb5a2c092ed75a5baf/pkg%2Foperator%2Fresource%2Fresourceapply%2Fapps.go#L107